app/src/androidTest/java/com/sdp/movemeet/map/GPSRecordingActivityTest.java

```diff
@@ -0,0 +1,168 @@
+package com.sdp.movemeet.map;
+
+import android.Manifest;
+import android.location.Location;
+import android.location.LocationManager;
+import android.os.Looper;
+
+import androidx.test.core.app.ActivityScenario;
+import androidx.test.ext.junit.rules.ActivityScenarioRule;
+import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;
+import androidx.test.rule.GrantPermissionRule;
+import androidx.test.uiautomator.By;
+import androidx.test.uiautomator.UiDevice;
+import androidx.test.uiautomator.Until;
+
+import com.google.android.gms.location.FusedLocationProviderClient;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.tasks.Task;
+import com.sdp.movemeet.R;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.stubbing.Answer;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+
+import static androidx.test.espresso.Espresso.onView;
+import static androidx.test.espresso.action.ViewActions.click;
+import static androidx.test.espresso.assertion.ViewAssertions.matches;
+import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static androidx.test.espresso.matcher.ViewMatchers.withId;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@RunWith(AndroidJUnit4ClassRunner.class)
+public class GPSRecordingActivityTest {
+
+    ActivityScenario scenario;
+
+    private static final double FAKE_LATITUDE = 30.00;
+    private static final double FAKE_LONGITUDE = 5.00;
+    private static final float FAKE_ACCURACY = 3.0f;
+    private Location fakeLocation;
+
+    @Rule
+    public GrantPermissionRule mRuntimePermissionRule = GrantPermissionRule.grant(Manifest.permission.ACCESS_FINE_LOCATION);
+
+    @Rule
+    public GrantPermissionRule CoarseLocationPermissionRule = GrantPermissionRule.grant(Manifest.permission.ACCESS_COARSE_LOCATION);
+
+
+    @Rule
+    public ActivityScenarioRule<GPSRecordingActivity> testRule = new ActivityScenarioRule<>(GPSRecordingActivity.class);
+
+    private UiDevice device;
+
+    private Task<Void> mockTask;
```

Comment:
Why do you use mock ?

Reply:
I'm mocking the `FusedLocationProviderClient` and injecting my own code when the `.requestLocationUpdates()` method is called. I had to create a mock Task<Void> instance to return something consistent with the method signature, after having run my own testing code.